File: CRM-Client/src/schemas/workflow.ts
import { z } from 'zod'

const WorkflowNodeSchema = z.object({
  id: z.string(),
  type: z.string(),
  position: z.object({ x: z.number(), y: z.number() }).passthrough(),
  config: z.record(z.unknown()),
}).passthrough()

const WorkflowEdgeSchema = z.object({
  id: z.string(),
  source: z.string(),
  target: z.string(),
}).passthrough()

export const WorkflowDslSchema = z.object({
  schema_version: z.literal(1),
  nodes: z.array(WorkflowNodeSchema),
  edges: z.array(WorkflowEdgeSchema),
}).passthrough()

export const WorkflowSummarySchema = z.object({
  id: z.number(),
  name: z.string(),
  description: z.string().nullable().optional(),
  status: z.enum(['draft', 'published', 'paused']),
  node_count: z.number().int().nonnegative(),
  created_time: z.string(),
  last_modified_time: z.string(),
}).passthrough()

export const WorkflowDetailSchema = WorkflowSummarySchema.extend({
  dsl: WorkflowDslSchema,
  created_by: z.number().nullable().optional(),
})

export type WorkflowNode = z.infer<typeof WorkflowNodeSchema>
export type WorkflowDsl = z.infer<typeof WorkflowDslSchema>
export type WorkflowSummary = z.infer<typeof WorkflowSummarySchema>
export type WorkflowDetail = z.infer<typeof WorkflowDetailSchema>

File: CRM-Server/app/schemas/workflow.py
"""Workflow schemas: Pydantic boundary for /v1/workflows."""

from __future__ import annotations

from datetime import datetime
from typing import Any, Literal, Optional

from pydantic import BaseModel, ConfigDict, Field

WorkflowStatusLiteral = Literal["draft", "published", "paused"]


class WorkflowDslSchema(BaseModel):
    """CRMWolf Workflow DSL（schema_version 1）。结构校验由 workflow_dsl 服务执行。"""

    model_config = ConfigDict(extra="allow")

    schema_version: int = Field(..., description="DSL 版本，当前为 1")
    nodes: list[dict[str, Any]] = Field(..., description="节点列表")
    edges: list[dict[str, Any]] = Field(default_factory=list, description="边列表")


class WorkflowCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=100, description="工作流名称")
    description: Optional[str] = Field(None, description="描述")
    dsl: WorkflowDslSchema


class WorkflowUpdate(BaseModel):
    name: str = Field(..., min_length=1, max_length=100, description="工作流名称")
    description: Optional[str] = Field(None, description="描述")
    dsl: WorkflowDslSchema
    expected_last_modified_time: datetime = Field(
        ..., description="乐观锁：客户端读取时的 last_modified_time"
    )


class WorkflowStatusUpdate(BaseModel):
    status: WorkflowStatusLiteral


class WorkflowSummary(BaseModel):
    id: int
    name: str
    description: Optional[str] = None
    status: WorkflowStatusLiteral
    node_count: int
    created_time: datetime
    last_modified_time: datetime


class WorkflowDetail(WorkflowSummary):
    dsl: WorkflowDslSchema
    created_by: Optional[int] = None
